pkg/virtual/initializingworkspaces/builder/caching_authorizer.go

```diff
@@ -0,0 +1,71 @@
+package builder
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	kcpkubernetesclientset "github.com/kcp-dev/client-go/kubernetes"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/initialization"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/authorization/delegated"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+	"github.com/kcp-dev/logicalcluster/v2"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/klog/v2"
+)
+
+func newCachingAuthorizer(client kcpkubernetesclientset.ClusterInterface) authorizer.Authorizer {
+	return &cachingAuthorizer{client: client, cache: make(map[string]authorizer.Authorizer)}
+}
+
+type cachingAuthorizer struct {
+	client kcpkubernetesclientset.ClusterInterface
+
+	cacheMu sync.RWMutex
+	cache   map[string]authorizer.Authorizer
```

Comment:
Would we prefer an LRU cache here instead? Would that make the caching authorizer struct a bit more portable between here and other places?

Reply:
This is probably an ok start? It doesn't remove entries when a workspace is deleted, but we could add that if we wanted.

Reply:
How would I know within the builder that a workspace has been deleted?

Reply:
We'd do like we've done in other places (e.g. quota) - spin up a controller that reacts to informer events, in this case deletions.

Reply:
Got it! That sounds good and can definitely do that, is that preferable to a basic LRU cache?

Reply:
I'd probably start with the LRU as that's less code. WDYT?

Reply:
pushed, ptal